dashboard/modules/reporter/reporter_agent.py

```diff
@@ -394,30 +397,66 @@ def _get_cpu_percent(in_k8s: bool):
     @staticmethod
     def _get_gpu_usage():
         global enable_gpu_usage_check
-        if gpustat is None or not enable_gpu_usage_check:
+        if not enable_gpu_usage_check:
             return []
         gpu_utilizations = []
-        gpus = []
         try:
-            gpus = gpustat.new_query().gpus
+            pynvml.nvmlInit()
+            num_gpus = pynvml.nvmlDeviceGetCount()
+            for i in range(num_gpus):
+                gpu_handle = pynvml.nvmlDeviceGetHandleByIndex(i)
+                utilization = None
+                try:
+                    utilization_info = pynvml.nvmlDeviceGetUtilizationRates(gpu_handle)
+                    utilization = int(utilization_info.gpu)
+                except pynvml.NVMLError as e:
+                    logger.debug(f"pynvml failed to retrieve GPU utilization: {e}")
+
+                # processes pids
+                processes_pids = None
+                try:
+                    nv_comp_processes = pynvml.nvmlDeviceGetComputeRunningProcesses(
```

Comment:
@wookayin actually we need to do the same thing as https://github.com/wookayin/gpustat/pull/143/ if we want to support older version of nvidia drivers?

Reply:
Yes, your understanding is correct. The current bundled version [12.535.133](https://github.com/wookayin/nvidia-ml-py/blob/12.535.133/pynvml.py (actually since 11.515.0) uses `nvmlDeviceGetGraphicsRunningProcesses_v3` API, which will raise "function not found error" for drivers < 510.39.

Similarily, the v2 API was introduced in nvidia-ml-py=11.450.129, requires NVIDIA 450.66+. Legacy drivers as low as R430 (which is pretty outdated at this point) will a pynvml version compatible with `nvidia-ml-py<11.450.129`.

So in order to support "old" NVIDIA drivers you'll need to do monkey-patch similar in wookayin/gpustat#143, or modify the `pynvml.py` bundled in ray in such a way that the function uses the v2 C function (or even the v1 function). If this is to be done without a third-party library such as gpustat or nvitop, it'd be quite bothersome.

Reply:
FYI `gpustat` requires NVIDIA 450.66 (August 2020) or higher (old versions of gpustat can work with R430 drivers though) due to `nvmlDeviceGetGraphicsRunningProcesses_v2` (v2) API.

BTW, MIG support is available only in 11.450.129+, so without a very cumbersome fallback mechanism which I don't think it's worth doing, the minimum driver version ray would want to require would be 450.66 at least, which I think is a reasonable requirement.

Reply:
So in order to support 450.66, we need to implement the fallback from v3 to v2. Is my understanding correct? What do you mean by `which I don't think it's worth doing`?

Reply:
Exactly. I meant it is not worth the effort to support drivers even lower than 450.66 (that does not have v2 process API). We'll need some compatibility layer (either fallback as in wookayin/gpustat#143 or manually modifying `pynvml.py`).

Reply:
@wookayin I guess another option is to downgrade the bundled version to one version below `11.515.0` that still users v2 API and also still has MIG support. I assume this will then work with all drivers 450.66+?

Reply:
Yes. I think that would be the easiest solution but make sure you test with the actual driver instance. Something like 11.495.46? Actually, I find that the v3 API was introduced in 11.510.69 not 11.515.0 (sorry my mistake).

You will find useful:
- https://github.com/wookayin/nvidia-ml-py/blame/main/pynvml.py
- https://github.com/wookayin/nvidia-ml-py/commits/main

Reply:
@wookayin do you know the main difference between v2 and v3 APIs? seems like for our use case, v2 will be enough, but is there other component we need to aware from v3?

Reply:
I think the main difference of the v3 API is the data structure on process detail. NVIDIA never properly documented this anywhere. But, I guess using v2 API should be fine for you, but I haven't really tested calling v2 API on recent NVIDIA drivers.

FYI there was [a huge mess in 535 driver versions](https://forums.developer.nvidia.com/t/nvml-12-535-43-02-breaks-backwards-compatibility/254999/17). In some broken versions v3 used a different data structure (`nvmlProcessInfo_v2_st` with *different* byte size), and suddenly removed & reverted that, etc. After everything is resolved (535.104.05+), the v3 API uses the same struct `nvmlProcessInfo_v2_st` as v2 API so I think they are basically the same.

Looping @XuehaiPan here, please feel free to correct if I said anything wrong.

Reply:
Agree with @wookayin. To support a wide range of NVIDIA driver verions without complex monkey patching, we'd better use `nvmlDeviceGet{Compute,Graphics}RunningProcesses_v2` APIs with `nvmlProcessInfo_v2` type.